context.R

```diff
@@ -67,6 +67,84 @@ context <- function(testcases={}, preExec={}) {
     )
 }
 
+
+contextWithParsing <- function(testcases={}, preExec={}) {
+    testcases <- substitute(testcases)
+
+    get_reporter()$start_context()
+    do_exit <- TRUE
+    on.exit({
+        if(do_exit) {
+            get_reporter()$end_context()
+        }
+    })
+
+    # remove unnamed list items and throw a warning
+    filtered_testcases <- testcases[!duplicated(names(testcases)) & names(testcases)!=""]
+    if(length(filtered_testcases) > length(testcases)){
+        get_reporter()$add_message("warning: There are duplicate names and/or unamed testcases found. Note that these won't be used for evaluation.")
+    }
+
+    # parse the student code into a named list linking the parsed codeblock names to codeblocks.
+    codeblocks <- list()
+    codeblock_name <- NULL
+    codeblock <- c()
+    for(line in read_lines(student_code)){
+        match <- str_match(line, "^###\\h*(.+[^\\h])\\h*###")[,2]
+        if(match %in% names(codeblocks)){
+            # duplicate template name
+            # not sure if i want to add a warning here
```

Comment:
A soft warning to the student feels appropriate.

Reply:
warning added